### PR TITLE
Fixing more bugs in emulator. 

### PR DESCRIPTION
**Pull request by weinbe58**
1. Detuning operator had the wrong sign
2. Rabi amplitude was missing a factor of 1/2
3. Added a [KS test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test#Two-sample_Kolmogorov%E2%80%93Smirnov_test) to check of the empirical distributions generated by bloqade and braket are statistically the same. 
4. Adding testss for emulator IR code generation.

### CODE REVIEW

File: tests/test_python_emulator.py [hunk 1]
```diff
@@ -159,3 +161,87 @@ def test_serialization():
     obj_str = dumps(batch)
     batch2 = loads(obj_str)
     assert isinstance(batch2, type(batch))
+
+
+def KS_test(
```
Comment (Roger-luo) [inline tests/test_python_emulator.py:166]: is there a package for this? or we are gonna maintain this

File: tests/test_python_emulator.py [hunk 2]
```diff
@@ -159,3 +161,87 @@ def test_serialization():
...
+    rhs_cdf = np.cumsum(rhs_cdf)
+
+    c_alpha = np.sqrt(-0.5 * np.log(alpha / 2))
+
+    assert np.linalg.norm(lhs_cdf - rhs_cdf, ord=np.inf) <= c_alpha * np.sqrt(
+        (m + n) / (m * n)
+    )
+
+
+def test_bloqade_against_braket():
```
Comment (Roger-luo) [inline tests/test_python_emulator.py:190]: can we get a rough estimate on how long this emulator test gonna take? so we don't keep CI running too long.

File: src/bloqade/codegen/common/is_hyperfine.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+class IsHyperfineSequence(AnalogCircuitVisitor):
+    def __init__(self):
+        self.is_hyperfine = False
+
+    def visit_analog_circuit(self, ast: analog_circuit.AnalogCircuit) -> Any:
+        self.visit(ast.sequence)
+
+    def visit_append_sequence(self, ast: sequence.Append) -> Any:
+        list(map(self.visit, ast.sequences))
```
Comment (Roger-luo) [inline src/bloqade/codegen/common/is_hyperfine.py:15]: coverage complains these visitors are not tested

**Comment by Roger-luo** [comment]
haven't looked into the detailed implementation, I'll do another review tmr

File: src/bloqade/codegen/common/is_hyperfine.py (hunk 3 above)
Comment (weinbe58) [reply src/bloqade/codegen/common/is_hyperfine.py:15]: we don't currently support appended sequences in the builder so I can just comment this out

File: tests/test_python_emulator.py (hunk 2 above)
Comment (weinbe58) [reply tests/test_python_emulator.py:190]: the total time for all the tests for the emulator takes about 25 seconds or so.

File: tests/test_python_emulator.py (hunk 1 above)
Comment (weinbe58) [reply tests/test_python_emulator.py:166]: looks like there is am implementatoin in scipy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ks_2samp.html#scipy.stats.ks_2samp

File: src/bloqade/emulate/ir/state_vector.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ class RabiOperator:
     phase: Optional[Callable[[float], float]] = None
 
     def dot(self, register: NDArray, time: float):
-        amplitude = self.amplitude(time)
+        amplitude = self.amplitude(time) / 2
```
Comment (Roger-luo) [inline src/bloqade/emulate/ir/state_vector.py:37]: can you explain why this is div by 2?

File: tests/test_python_emulator.py (hunk 2 above)
Comment (Roger-luo) [reply tests/test_python_emulator.py:184]: that's kinda ok, might be a bit long, is it possible to run a smaller test? (e.g 4 atoms or even 2 atoms?)

File: src/bloqade/emulate/ir/state_vector.py (hunk 4 above)
Comment (weinbe58) [reply src/bloqade/emulate/ir/state_vector.py:37]: that's how we define the Ryberg Hamiltonian

Comment (weinbe58) [reply src/bloqade/emulate/ir/state_vector.py:37]: Remember this guy? https://github.com/QuEraComputing/Bloqade.jl/blob/c219ebfbfc750e260e52c9b57984582b4e71dfcf/lib/BloqadeExpr/src/types.jl#L689C8-L689C16

File: tests/test_python_emulator.py (hunk 2 above)
Comment (weinbe58) [reply tests/test_python_emulator.py:178]: OK I reduce time time to ~7 seconds

File: tests/test_emulator_codegen.py [hunk 5]
```diff
@@ -111,3 +113,269 @@ def test_codegen_detuning_and_rabi():
...
+    drives = {LevelCoupling.Rydberg: rydberg_drive}
+
+    expected_emulator_ir = EmulatorProgram(
+        register=geometry, duration=1.0, pulses=drives
+    )
+
+    emulator_ir = EmulatorProgramCodeGen(assignments=assignments).emit(program)
+
+    print(emulator_ir)
+    print(expected_emulator_ir)
```
Comment (Roger-luo) [inline tests/test_emulator_codegen.py:244]: this is something forgot to delete?

**Review by Roger-luo — APPROVED** [review]

File: tests/test_emulator_codegen.py (hunk 5 above)
Comment (weinbe58) [reply tests/test_emulator_codegen.py:244]: its in a test so I don't think this is an issue.

Comment (weinbe58) [reply tests/test_emulator_codegen.py:244]: its in the tests so I don't think this is an issue

**Pull request merged**